### PR TITLE
[CALCITE-3381] When converting rel to bigquery sql dialect, types should be converted to BigQuery's data types

### PR DESCRIPTION
**Pull request by amaliujia**
The mapping is as the following, and the left side is Calcite types and right side is BigQuery data types:

VARCHAR -> STRING
VARBINARY -> BYTES
BIGINT -> INT64
DECIMAL -> NUMERIC
DOUBLE -> FLOAT64
BOOLEAN -> BOOL
DATE -> DATE
TIME -> TIME
TIMESTAMP -> TIMESTAMP

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -132,6 +138,55 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+        return new SqlDataTypeSpec(typeNameSpec, SqlParserPos.ZERO);
+      case TIMESTAMP:
+        typeNameSpec = new SqlUserDefinedTypeNameSpec(
+                new SqlIdentifier("TIMESTAMP", SqlParserPos.ZERO), SqlParserPos.ZERO);
+        return new SqlDataTypeSpec(typeNameSpec, SqlParserPos.ZERO);
+      }
+    }
+    return super.getCastSpec(type);
+  }
+
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Consider adding a helper method to reduce each of those 3 line blocks to a single line. Otherwise, looks good.

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Good idea. Done.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -132,6 +138,43 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
     }
   }
 
+  /** BigQuery data type reference:
+   * <a href="https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types">
+   * Bigquery Standard SQL Data Types</a>
+   */
+  @Override public SqlNode getCastSpec(final RelDataType type) {
+    if (type instanceof BasicSqlType) {
+      SqlUserDefinedTypeNameSpec typeNameSpec;
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: This can be removed.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 3]
```diff
@@ -132,6 +138,43 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
...
+      case TIME:
+        return createSqlDataTypeSpecByName("TIME");
+      case TIMESTAMP:
+        return createSqlDataTypeSpecByName("TIMESTAMP");
+      }
+    }
+    return super.getCastSpec(type);
+  }
+
+  private static SqlDataTypeSpec createSqlDataTypeSpecByName(String identifierName) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: why static?

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 4]
```diff
@@ -143,7 +186,6 @@ public BigQuerySqlDialect(SqlDialect.Context context) {
 
   private static final SqlSetOperator INTERSECT_DISTINCT =
       new SqlSetOperator("INTERSECT DISTINCT", SqlKind.INTERSECT, 18, false);
-
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: undo this change

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 3 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I usually mark a helper function as a static if it does not access class member. Is there an option on static/non-static in this case?

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: oops. Removed.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 4 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Done. I think I should check `git diff` more frequently.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 3 above)
Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Unless it is a utility class, or the method is called by another static method, I think it should be non-static method. e.g. `Aggregate.isPredicate`

Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Make it as non-static. thanks for the example of Calcite practice.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: There isn't really a standard practice. Note that a few lines above `Aggregate.isPredicate`, which is static, is `isNotGrandTotal` which is non-static.

In my opinion, if a method is `private` it doesn't really matter. Once it becomes `public` or `protected` it can be overridden, referenced from places that can't see the implementation, etc. So that's when some serious thought has to take place.
